draft-thomson-webpush-vapid.md

```diff
@@ -132,211 +134,279 @@ The words "MUST", "MUST NOT", "SHOULD", and "MAY" are used in this document.
 It's not shouting, when they are capitalized, they have the special meaning
 described in [RFC2119].
 
-The terms "push message", "push service", "application server", and "user agent"
-are used as defined in [I-D.ietf-webpush-protocol]
+The terms "push message", "push service", "push subscription", "application
+server", and "user agent" are used as defined in [I-D.ietf-webpush-protocol].
 
 
-# Self-Identification Alternatives
+# Application Server Self-Identification {#jwt}
 
-Several options have been proposed, here are some of the more concrete options.
+Application servers SHOULD generate and maintain a signing key pair usable with
+elliptic curve digital signature (ECDSA) over the P-256 curve [FIPS186].  Use of
+this key when sending push messages establishes a continuous identity for the
+application server.
 
+When requesting delivery of a push message, the application includes a JSON Web
+Token (JWT) [RFC7519], signed using its signing key.  The token includes a
+number of claims as follows:
 
-## Certificates
+ * An "aud" (Audience) claim in the token MUST include the unicode serialization
+   of the origin (Section 6.1 of [RFC6454]) of the push resource URL.  This
+   binds the token to a specific push service.  This ensures that the token is
+   reusable for all push resource URLs that share the same origin.
 
-A push service that supports application server self-identification requests a
-client certificate from application servers.  The client certificate is
-requested during the TLS [RFC5246] handshake.
+ * An "exp" (Expiry) claim MUST be included with the time after which the token
+   expires.  This limits the time that a token over which a token is valid.  An
+   "exp" claim MUST NOT be more than 24 hours from the time of the request.
 
-An application server that does not have a client certificate offers no
-certificate in response; an application server that wishes to self-identify
-includes a certificate.
+This JWT is included in an Authorization header field, using an auth-scheme of
+"WebPush".  A push service MAY reject a request with a 401 (Unauthorized) status
```

Comment:
SHOULD implies that the push service SHOULD also validate the token on every request.  I think that it would be easier in some cases to defer validation.  MAY has the same effect on validity; application servers won't risk sending a bad token if there is a chance that the server will reject it.